core/src/main/java/org/infinispan/marshall/core/internal/ExternalJBossMarshaller.java

```diff
@@ -0,0 +1,226 @@
+package org.infinispan.marshall.core.internal;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.io.OutputStream;
+
+import org.infinispan.commons.io.ByteBuffer;
+import org.infinispan.commons.marshall.BufferSizePredictor;
+import org.infinispan.commons.marshall.Externalizer;
+import org.infinispan.commons.marshall.MarshallableTypeHints;
+import org.infinispan.commons.marshall.StreamingMarshaller;
+import org.infinispan.configuration.global.GlobalConfiguration;
+import org.infinispan.marshall.core.JBossMarshaller;
+import org.jboss.marshalling.ByteInput;
+import org.jboss.marshalling.ByteOutput;
+import org.jboss.marshalling.ObjectTable;
+import org.jboss.marshalling.Unmarshaller;
+
+final class ExternalJBossMarshaller implements StreamingMarshaller {
+
+   final MarshallableTypeHints marshallableTypeHints = new MarshallableTypeHints();
+   final JBossMarshaller marshaller;
+
+   public ExternalJBossMarshaller(InternalExternalizerTable externalizers, GlobalConfiguration globalCfg) {
+      this.marshaller = new InternalJBossMarshaller(externalizers, globalCfg);
+   }
+
+   @Override
+   public void objectToObjectStream(Object obj, ObjectOutput out) throws IOException {
+      assert ExternallyMarshallable.isAllowed(obj) : "Check support for: " + obj.getClass();
+
+      if (out instanceof PositionalBuffer.Output) {
+         PositionalBuffer.Output posOut = (PositionalBuffer.Output) out;
+
+         BufferSizePredictor sizePredictor = marshallableTypeHints
+               .getBufferSizePredictor(obj.getClass());
+         int estimatedSize = sizePredictor.nextSize(obj);
+
+         int beforePos = posOut.savePosition();
+
+         ObjectOutput jbossOut = marshaller.startObjectOutput(
+               new JBossByteOutput(out), false, estimatedSize);
+         try {
+            jbossOut.writeObject(obj);
+         } finally {
+            marshaller.finishObjectOutput(jbossOut);
+         }
+
+         int afterPos = posOut.writePosition(beforePos);
+         sizePredictor.recordSize(afterPos - beforePos);
+      } else {
+         throw new IOException("External JBoss Marshalling requires position tracking object output");
+      }
+   }
+
+   @Override
+   public Object objectFromObjectStream(ObjectInput in) throws IOException, ClassNotFoundException {
+      if (in instanceof PositionalBuffer.Input) {
+         PositionalBuffer.Input posIn = (PositionalBuffer.Input) in;
+
+         int pos = in.readInt();
```

Comment:
I'm not sure you really need this. `limit` and `position` are `protected` in `SimpleDataInput`, so you can add a method in our `ExtendedRiverUnmarshaller` to return `position - limit` and rewind by that amount.